### PR TITLE
evm/vm: reintroduce evm interface

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Addresses #2867 

Will await an approval that this is OK before continuing.

Should we also introduce to use this interface in client?

### CODE REVIEW

**Comment by holgerd77** [comment]
So, just to give some feedback here: for me this looks pretty healthy, generally make things a lot more transparent to see at one place what is actually exposed by the EVM (and used). And e.g. the journal then *is* used anyhow - being it implicit or named/stated in the interface - that's just how things are implemented right now and we just also do not want to change any more at this point (right?).

So, tldr;: I would be a fan of adding this here.

**Comment by jochem-brouwer** [comment]
> So, just to give some feedback here: for me this looks pretty healthy, generally make things a lot more transparent to see at one unifed place what is actually exposed by the EVM (and used). And e.g. the journal then _is_ used anyhow - being it implicit or named/stated in the interface - that's just how things are implemented right now and we just also do not want to change any more at this point (right?).
> 
> So, tldr;: I would be a fan of adding this here.

Ok, I wanted to just get this approval in order to continue, thanks :smile: 

Will finish today!

**Comment by holgerd77** [comment]
Updated this via UI

**Comment by holgerd77** [comment]
Client tests failing here.

File: packages/vm/src/vm.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ import type {
 } from './types.js'
 import type { BlockchainInterface } from '@ethereumjs/blockchain'
 import type { EVMStateManagerInterface } from '@ethereumjs/common'
+import type { EVMInterface } from '@ethereumjs/evm/dist/cjs/types.js'
```
Comment (holgerd77) [inline packages/vm/src/vm.ts:23]: These imports are still a total no-go for merging, I guess that's obvious?

Comment (jochem-brouwer) [reply packages/vm/src/vm.ts:23]: Huh, how did this get introduced? :thinking: I think VSCode auto-imported this

**Comment by jochem-brouwer** [comment]
When I got the approval I saw this comment on this PR (see first message): 

> Will await an approval that this is OK before continuing.

I was a bit confused why I had this comment, but now if I review my own PR I see many TODOs. This is also why this PR looks dirty (CC @holgerd77 )

File: packages/vm/src/vm.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/vm/src/vm.ts:23]: Nvm see my new comment on the PR, PR was not yet ready to be reviewed (sorry for marking it as ready-to-review)

**Comment by jochem-brouwer** [comment]
Have changed a method name (`reportAccessList` -> `startReportingAccessList`) in EVM journal interface since the name was misleading (not super related to this PR but this is an improvement IMO)

File: packages/vm/src/vm.ts [hunk 2]
```diff
@@ -231,7 +232,7 @@ export class VM {
       blockchain,
       stateManager,
     }
-    const evmCopy = new EVM(evmOpts)
+    const evmCopy = new EVM(evmOpts) // TODO fixme (should copy the EVMInterface, not default EVM)
```
Comment (gabrocheleau) [inline packages/vm/src/vm.ts:235]: Imo we could do this in this PR by just adding a `copy` method to the EVMInterface, and implemting it in EVM in the way you've done here. Then replace this line with `evmCopy = evm.copy()`. What do you think?

File: packages/evm/src/types.ts [hunk 3]
```diff
@@ -134,6 +136,26 @@ export type EVMEvents = {
...
+    deleteAccount(address: Address): Promise<void>
+    accessList?: Map<string, Set<string>>
+    addAlwaysWarmAddress(address: string, addToAccessList?: boolean): void
+    addAlwaysWarmSlot(address: string, slot: string, addToAccessList?: boolean): void
+    startReportingAccessList(): void
+  }
+  stateManager: EVMStateManagerInterface
+  precompiles: Map<string, PrecompileFunc>
+  runCall(opts: EVMRunCallOpts): Promise<EVMResult>
+  events?: AsyncEventEmitter<EVMEvents>
```
Comment (gabrocheleau) [inline packages/evm/src/types.ts:156]: Why are we making events optional? Could there be a valid EVM implementation that does not emit events?

**Review by gabrocheleau — COMMENTED** [review]
Looks good, left 2 comments.

File: packages/evm/src/types.ts (hunk 3 above)
Comment (jochem-brouwer) [reply packages/evm/src/types.ts:156]: It is not super-necessary to make EVM run. If we make this non-optional we need to have "external" EVMs import our AsyncEventEmitter and implement the events it defines

File: packages/vm/src/vm.ts (hunk 2 above)
Comment (jochem-brouwer) [reply packages/vm/src/vm.ts:235]: I think we should keep this as a TODO for now (?)

Comment (gabrocheleau) [reply packages/vm/src/vm.ts:235]: Ah sure yes, I thought this was a simple implementation but since it's more complicated than that let's leave for another PR :)

File: packages/vm/test/tester/runners/GeneralStateTestsRunner.ts [hunk 4]
```diff
@@ -129,7 +129,7 @@ async function runTestCase(options: any, testData: any, t: tape.Test) {
       const block = makeBlockFromEnv(testData.env, { common })
 
       if (options.jsontrace === true) {
-        vm.evm.events.on('step', stepHandler)
+        vm.evm.events!.on('step', stepHandler)
```
Comment (gabrocheleau) [inline packages/vm/test/tester/runners/GeneralStateTestsRunner.ts:132]: Would it make sense in these two instances to check if `vm.evm.events` is defined, and to throw if not?

Comment (jochem-brouwer) [reply packages/vm/test/tester/runners/GeneralStateTestsRunner.ts:132]: No, in tests we are sure we use default EVM, so we explicitly are sure that we use `EVM` and thus the events are defined :smile:

Comment (gabrocheleau) [reply packages/vm/test/tester/runners/GeneralStateTestsRunner.ts:132]: Ok sure!

**Review by gabrocheleau — APPROVED** [review]

**Comment by holgerd77** [comment]
Have updated this via UI

**Comment by holgerd77** [comment]
Maybe just to throw in the ring: do we want to keep this interface on EVM or also move to Common?

Side note: in case we move to Common I would be a fan of a slightly different structure there and do folder "interfaces/" and then have a dedicated named file for every major interface. This just scales better and is more transparent and very well worth to have an own directory for this, these interfaces now play a very central role in our monoprepo and can very well get a visible place.

**Comment by jochem-brouwer** [comment]
I indeed thought of moving this to Common and I think regarding cleani-ness this is great :smile:

**Comment by holgerd77** [comment]
Ok, can you maybe tackle directly today or tomorrow? Would you also be ok with this slightly adopted folder structure (and could you integrate as well)? So I would then name the files after the packages for simplicity: so e.g. `interfaces/evm.ts`.
